test/e2e/pull_chunked_test.go

```diff
@@ -0,0 +1,419 @@
+
+					// Test fresh pull
+					c.fresh.testPull(c.img, c.insecureStorage, &lastPullOutput)
+
+					podmanTest.PodmanExitCleanly("--pull-option=enable_partial_images=true", fmt.Sprintf("--pull-option=insecure_allow_unpredictable_image_contents=%v", c.insecureStorage),
+						"pull", "-q", "--tls-verify=false", chunkedNormal.registryRef)
+
+					// Test pull after chunked layers are already known, to trigger the layer reuse code
+					c.reuse.testPull(c.img, c.insecureStorage, &lastPullOutput)
+
+					pullChunkedStopRegistry()
+				})
+			}
+		})
+	})
+}
+
+const pullChunkedRegistryPort = "5013"
+
+// pullChunkedStartRegistry creates a registry listening at pullChunkedRegistryPort within the current Podman environment.
+func pullChunkedStartRegistry() {
+	podmanTest.PodmanExitCleanly("run", "-d", "--name", "registry", "--rm", "-p", pullChunkedRegistryPort+":5000", "-e", "REGISTRY_COMPATIBILITY_SCHEMA1_ENABLED=true", REGISTRY_IMAGE, "/entrypoint.sh", "/etc/docker/registry/config.yml")
+	if !WaitContainerReady(podmanTest, "registry", "listening on", 20, 1) {
+		Fail("Cannot start docker registry.")
+	}
+}
+
+// pullChunkedStopRegistry stops a registry started by pullChunkedStartRegistry.
+func pullChunkedStopRegistry() {
+	podmanTest.PodmanExitCleanly("stop", "registry")
+}
+
+// pullChunkedTestImage centralizes data about a single test image in pullChunkedTests.
+type pullChunkedTestImage struct {
+	registryRef, dirPath string
+	configDigest         digest.Digest // "" for a schema1 image
+}
+
+// push copies the image from dirPath to registryRef.
+func (img *pullChunkedTestImage) push() {
+	skopeo := SystemExec("skopeo", []string{"copy", "-q", fmt.Sprintf("--preserve-digests=%v", img.configDigest != ""), "--all", "--dest-tls-verify=false", "dir:" + img.dirPath, img.registryRef})
+	skopeo.WaitWithDefaultTimeout()
+	Expect(skopeo).Should(ExitCleanly())
+}
+
+// pullChunkedExpectations records the expected output of a single "podman pull" command.
+type pullChunkedExpectation struct {
+	success []string // Expected debug log strings; should succeed if != nil
+	failure []string // Expected debug log strings; should fail if != nil
+}
+
+// testPull performs one pull
+// It replaces *lastPullOutput with the output of the current command
+func (expectation *pullChunkedExpectation) testPull(image *pullChunkedTestImage, insecureStorage bool, lastPullOutput *bytes.Buffer) {
+	lastPullOutput.Reset()
+	session := podmanTest.PodmanWithOptions(PodmanExecOptions{
+		FullOutputWriter: lastPullOutput,
+	}, "--log-level=debug", "--pull-option=enable_partial_images=true", fmt.Sprintf("--pull-option=insecure_allow_unpredictable_image_contents=%v", insecureStorage),
+		"pull", "--tls-verify=false", image.registryRef)
+	session.WaitWithDefaultTimeout()
+	log := session.ErrorToString()
+	if expectation.success != nil {
+		Expect(session).Should(Exit(0))
+		for _, s := range expectation.success {
+			Expect(regexp.MatchString(".*"+s+".*", log)).To(BeTrue(), s)
```

Comment:
note these expected(flase).To(BeTrue()) lead to poor error messages
you should use something like `Expect(string).To(MatchRegexp("regex"))` instead

sorry for not noticing this earlier

Reply:
Yes; using the idiomatic way would trigger posting the _whole_ debug log of that Podman command into that error structure. “Expected … $theWholeDebugLog … to match $regexp”. E.g. the stack trace would then be rather far from the failure header.